File: lc-python/src/lc_0234_palindrome_linked_list.py
import unittest
from typing import Optional

from shared.linked_list import ListNode, list_to_list_node


def is_palindrome0(head: Optional[ListNode]) -> bool:
    """
    Time complexity: O(n)
    Space complexity: O(n)
    """

    def equals(
        n1: Optional[ListNode],
        n2: Optional[ListNode],
    ) -> bool:
        if n1 is None and n2 is None:
            return True
        return n1.val == n2.val and equals(n1.next, n2.next)

    def reverse(
        n1: Optional[ListNode],
        n2: Optional[ListNode] = None,
    ) -> Optional[ListNode]:
        # does NOT mutate the original list
        # object creation in python is REALLY REALLY REALLY slow...
        # the statement `ListNode(n1.val, n2)` takes relatively forever...
        if n1 is None:
            return n2
        return reverse(n1.next, ListNode(n1.val, n2))

    return equals(head, reverse(head))


def is_palindrome1(head: Optional[ListNode]) -> bool:
    """
    Time complexity: O(n)
    Space complexity: O(n)
    """

    stack = []
    slow = head
    fast = head
    while slow and fast:
        if fast.next is None:
            # list has odd-length, skips middle element
            slow = slow.next
            break
        fast = fast.next.next

        stack.append(slow.val)
        slow = slow.next

    while slow:
        n1 = slow.val
        n2 = stack.pop()
        if n1 != n2:
            return False
        slow = slow.next

    return True


def is_palindrome2(head: Optional[ListNode]) -> bool:
    """
    This approach mutates the original list.

    Time complexity: O(n)
    Space complexity: O(1)
    """

    slow = head
    fast = head
    rev = None
    while slow and fast:
        if fast.next is None:
            # list has odd-length, skips middle element
            slow = slow.next
            break
        fast = fast.next.next

        # head will point to the tail once done
        # append to rev
        tmp = slow.next
        slow.next = rev
        rev = slow
        slow = tmp

    while rev:
        if slow.val != rev.val:
            return False
        slow = slow.next
        rev = rev.next

    return True


class TestPalindromeLinkedList(unittest.TestCase):
    @staticmethod
    def parse_input():
        data = [
            [[1, 2, 2, 1], True],
            [[1, 2, 3, 2, 1], True],
            [[1, 2], False],
            [[1], True],
        ]
        return data

    def test_palindrome_linked_list0(self):
        test_cases = self.parse_input()
        for tc in test_cases:
            h = list_to_list_node(tc[0])
            expected = tc[1]

            actual = is_palindrome0(h)

            self.assertEqual(actual, expected)

    def test_palindrome_linked_list1(self):
        test_cases = self.parse_input()
        for tc in test_cases:
            h = list_to_list_node(tc[0])
            expected = tc[1]

            actual = is_palindrome1(h)

            self.assertEqual(actual, expected)

    def test_palindrome_linked_list2(self):
        test_cases = self.parse_input()
        for tc in test_cases:
            h = list_to_list_node(tc[0])
            expected = tc[1]

            actual = is_palindrome2(h)

            self.assertEqual(actual, expected)


if __name__ == "__main__":
    unittest.main()

File: lc-typescript/src/solutions/lc_0234_palindrome_linked_list.ts
import { assertEquals } from "https://deno.land/std@0.224.0/assert/assert_equals.ts";
import { ListNode, array_to_linked_list } from "../shared/utils.ts";

function is_palindrome(head: ListNode | null): boolean {
  let rev: ListNode | null = null;
  let slow: ListNode | null = head;
  let fast: ListNode | null = head;
  while (slow != null && fast != null) {
    if (fast.next == null) {
      // list has odd-length, skips middle element
      slow = slow.next;
      break;
    }
    fast = fast.next.next;

    // build reverse
    const tmp: ListNode | null = slow.next;
    slow.next = rev;
    rev = slow;
    slow = tmp;
  }

  while (slow != null && rev != null) {
    if (slow.val !== rev.val) {
      return false;
    }
    rev = rev.next;
    slow = slow.next;
  }

  return true;
}

Deno.test("0234 Palindrom Linked List", () => {
  const test_cases: [number[], boolean][] = [
    [[1, 2, 2, 1], true],
    [[1, 2, 3, 2, 1], true],
    [[1, 2], false],
    [[1], true],
  ];

  for (let i = 0; i < test_cases.length; i += 1) {
    const test_case = test_cases[i];

    const head = array_to_linked_list(test_case[0]);
    const expected = test_case[1];

    const actual = is_palindrome(head);
    assertEquals(actual, expected, `test case: ${i}`);
  }
});
